File: db4oj.tests/src/com/db4o/db4ounit/common/fieldindex/RuntimeFieldIndexTestCase.java
/* Copyright (C) 2007   Versant Inc.   http://www.db4o.com */
package com.db4o.db4ounit.common.fieldindex;

import com.db4o.*;
import com.db4o.ext.*;
import com.db4o.query.*;

import db4ounit.*;
import db4ounit.extensions.*;
import db4ounit.extensions.fixtures.*;

public class RuntimeFieldIndexTestCase extends AbstractDb4oTestCase implements OptOutMultiSession {
	
	private static final String FIELDNAME = "_id";

	public static class Data {
		public int _id;

		public Data(int id) {
			_id = id;
		}		
	}
	
	protected void store() throws Exception {
		for(int i=1; i <= 3; i++) {
			store(new Data(i));
		}
	}
	
	public void testCreateIndexAtRuntime() {
		StoredField field = storedField();
		Assert.isFalse(field.hasIndex());
		field.createIndex();
		Assert.isTrue(field.hasIndex());
		assertQuery();
		field.createIndex(); // ensure that second call is ignored
	}

	private void assertQuery() {
		Query query = newQuery(Data.class);
		query.descend(FIELDNAME).constrain(new Integer(2));
		ObjectSet result = query.execute();
		Assert.areEqual(1, result.size());
	}
	
	public void testDropIndex(){
		StoredField field = storedField();
		field.createIndex();
		assertQuery();
		field.dropIndex();
		Assert.isFalse(field.hasIndex());
		assertQuery();
	}

	private StoredField storedField() {
		return db().storedClass(Data.class).storedField(FIELDNAME,null);
	}

}

File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/Common/Fieldindex/RuntimeFieldIndexTestCase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using Db4oUnit;
using Db4oUnit.Extensions;
using Db4oUnit.Extensions.Fixtures;
using Db4objects.Db4o;
using Db4objects.Db4o.Ext;
using Db4objects.Db4o.Query;
using Db4objects.Db4o.Tests.Common.Fieldindex;

namespace Db4objects.Db4o.Tests.Common.Fieldindex
{
	public class RuntimeFieldIndexTestCase : AbstractDb4oTestCase, IOptOutMultiSession
	{
		private static readonly string Fieldname = "_id";

		public class Data
		{
			public int _id;

			public Data(int id)
			{
				_id = id;
			}
		}

		/// <exception cref="System.Exception"></exception>
		protected override void Store()
		{
			for (int i = 1; i <= 3; i++)
			{
				Store(new RuntimeFieldIndexTestCase.Data(i));
			}
		}

		public virtual void TestCreateIndexAtRuntime()
		{
			IStoredField field = StoredField();
			Assert.IsFalse(field.HasIndex());
			field.CreateIndex();
			Assert.IsTrue(field.HasIndex());
			AssertQuery();
			field.CreateIndex();
		}

		// ensure that second call is ignored
		private void AssertQuery()
		{
			IQuery query = NewQuery(typeof(RuntimeFieldIndexTestCase.Data));
			query.Descend(Fieldname).Constrain(2);
			IObjectSet result = query.Execute();
			Assert.AreEqual(1, result.Count);
		}

		public virtual void TestDropIndex()
		{
			IStoredField field = StoredField();
			field.CreateIndex();
			AssertQuery();
			field.DropIndex();
			Assert.IsFalse(field.HasIndex());
			AssertQuery();
		}

		private IStoredField StoredField()
		{
			return Db().StoredClass(typeof(RuntimeFieldIndexTestCase.Data)).StoredField(Fieldname
				, null);
		}
	}
}
